clients/cli/src/prover.rs

```diff
@@ -7,132 +7,251 @@ use crate::orchestrator_client::OrchestratorClient;
-            return Err(e.into());
-        }
-    };
-    let proof_hash = format!("{:x}", Keccak256::digest(&proof_bytes));
-
-    println!("Submitting ZK proof to Nexus Orchestrator...");
-    if let Err(e) = client
-        .submit_proof(node_id, &proof_hash, proof_bytes)
-        .await{
-            error!("Failed to submit proof: {}", e);
-            return Err(e);
-        }
-    
+
+    // Run the prover in the custom thread pool
+    let (view, proof) = pool.install(|| {
+        let prover = Stwo::<Local>::new_from_file(&elf_file_path)
+            .map_err(|e| ProverError::from(format!("Failed to load guest program: {}", e)))?;
+
+        println!("Creating ZK proof with inputs...");
+        prover
+            .prove_with_input::<(), u32>(&(), &public_input)
+            .map_err(|e| ProverError::from(format!("Failed to run prover: {}", e)))
+    })?;
+
+    let code = view.exit_code().map(|u| u as i32).unwrap_or_else(|_err| {
+        // We can't use await here, so just return -1 and let the error be handled below
+        -1
+    });
+
+    if code != 0 {
+        error!("Unexpected exit code: {}", code);
+        return Err(Box::new(ProverError::from(format!(
+            "Unexpected exit code: {}",
+            code
+        ))));
+    }
+
+    let proof_bytes = serde_json::to_vec(&proof)
+        .map_err(ProverError::from)?;
+    let _proof_hash = format!("{:x}", Keccak256::digest(&proof_bytes));
+
+    let _ = client
+        .submit_proof(
+            node_id,
+            &_proof_hash,
+            proof_bytes,
+            proof_task.task_id as u64,
+        )
+        .await;
+
     println!("{}", "ZK proof successfully submitted".green());
     Ok(())
 }
 
-fn anonymous_proving() -> Result<(), Box<dyn std::error::Error>> {
-    // 1. Instead of fetching the proof task from the orchestrator, we will use hardcoded input program and values
-
-    // The 10th term of the Fibonacci sequence is 55
-    let public_input: u32 = 9;
-    println!("Compiling guest program...");
-    let elf_file_path = std::path::Path::new(env!("CARGO_MANIFEST_DIR"))
-        .join("assets")
-        .join("fib_input");
+fn anonymous_proving(speed: &crate::ProvingSpeed) -> Result<(), Box<dyn StdError + Send + Sync>> {
+    // Set thread count based on speed
+    let num_threads = match speed {
+        crate::ProvingSpeed::Low => {
+            let total_cores = thread::available_parallelism().map_or(1, |n| n.get());
+            (total_cores + 3) / 4 // Use 25% of cores, rounded up
+        }
+        crate::ProvingSpeed::Medium => {
+            let total_cores = thread::available_parallelism().map_or(1, |n| n.get());
+            (total_cores + 1) / 2 // Use 50% of cores, rounded up
+        }
+        crate::ProvingSpeed::High => {
+            let total_cores = thread::available_parallelism().map_or(1, |n| n.get());
+            (total_cores * 3 + 3) / 4 // Use 75% of cores, rounded up
+        }
+    };
```

Comment:
I think this file might be cleaner if we do the speed -> num_threads calculations before calling the `anonymous_prooving`/`authenticated_prooving` methods.

You can do a TryInto I suppose.